### DISCUSSION TITLE
Zoom & rotate animation

### DISCUSSION THREAD

**Discussion by Sil68** [ImageMagick/ImageMagick#4443] [Help]
Hi,

I'd like to zoom into a given image, starting from a given point, and with each zoom iteration, also rotate the image a bit, while at the same time maintaining the overall image size.

I had a look at various corresponding entries on The Net, as well as at the -distort SRT command, but I'm afraid I'm failing at making head and tails out of this.

Any help would be more than appreciated.

My environment comprises of a Mac mini M1 sporting macOS Monterey, and ImageMagick 7.1.0-12 Q16-HDRI arm 2021-10-25.

-- 
Sil68

**Comment by snibgo** [comment]
See [Animation with SRT](http://im.snibgo.com/animsrt.htm). Examples there are for IM v6 and Windows BAT syntax, but easily translatable to IM v7 and bash.

**Reply by Sil68** [reply]
Thanks, that's one of the resources I have stumbled across. I first tried to execute each of the steps described in there manually, then trying to convert the bat into a shell script, but unfortunately I ended up with a ton of error messages reading

"convert: invalid argument for option ScaleRotateTranslate : 'Zero Scale Given' @ error/distort.c/GenerateCoefficients/811."

**Reply by fmw42** [reply]
Read the documentation. Only certain combinations of options can be used.

**Reply by snibgo** [reply]
> then trying to convert the bat into a shell script, but unfortunately I ended up with a ton of error messages reading

If you don't show your bash script, we can only guess at what is wrong.

**Reply by Sil68** [reply]
After some try-out I managed pulling this one

```
in_width=3000
in_height=4000

out_width=$in_width
out_height=$in_height

centre_x=`perl -e "printf(\"%0.9f\", $in_width / 2)"`
centre_y=`perl -e "printf(\"%0.9f\", $in_height / 2)"`

in_x=2728
in_y=3080

out_x=`perl -e "printf(\"%0.9f\", $in_width / 2)"`
out_y=`perl -e "printf(\"%0.9f\", $in_height / 2)"`

frames=50

zoom=1.1
scale=1
angle=37

d_move_x=`perl -e "printf(\"%0.9f\", ($in_x - $centre_x) / $frames)"`
d_move_y=`perl -e "printf(\"%0.9f\", ($in_y - $centre_y) / $frames)"`

d_in_x=3
d_in_y=0

d_out_x=3
d_out_y=1.5

d_scale=1.1
d_angle=1

convert07() {
	# $1 ... label
	# $2 ... input/source file
	# $3 ... output/destination file

	convert \
		-loop 0 -delay 20 \
		$2 \
		-duplicate $frames \
		-matte -virtual-pixel transparent \
		-distort SRT \
			"%[fx:$centre_x],%[fx:$centre_y] \
			 %[fx:$scale * pow($d_scale, t)] \
			 %[fx:($angle / $frames) * t] \
			 %[fx:$centre_x - (($in_x - $centre_x) / $frames) * t], \
			 %[fx:$centre_y - (($in_y - $centre_y) / $frames) * t]" \
		+repage \
		$3

} # convert07

convert07 "test:" test.png transform07.mp4
```

The image (3000x4000 px) gets rotated, scaled, and moved from the bottom right to the top left (but only if 4 is being used instead of $frames).

**Comment by fmw42** [comment]
-distort SRT "ix,iy scale rotation ox,oy" with the minus (-distort) always keeps the output to the same size as the input. Arguments can be computed in addition to be specifically set.

Rotates are always at the origin. So you use ix,iy to specify the rotation point such that that point will be moved to the origin. Then the image rotated and scaled. Then it will move the point at the origin to the specified ox,oy.

So if you want to rotate and scale at a point and keep that same location from input to output, set ix,iy = ox,oy. For example in Unix syntax for IM 6 (Linux, Mac OSX or Windows 10 Unix):

Input:

![lena](https://user-images.githubusercontent.com/13458218/140454152-0b9f2f4e-b747-4d1a-ab80-6966a11ee338.jpg)

Rotate and Zoom about 128,128
```
ix=128
iy=128
ox=128
oy=128
zoom=1.1
angle=2
convert lena.jpg -distort SRT "$ix,$iy $zoom $angle $ox,$oy" x1.jpg

```

![x1](https://user-images.githubusercontent.com/13458218/140454308-9f72ff80-fe0b-4b4c-91c0-9266b14901ad.jpg)

or

Rotate and zoom about 64,64
```
ix=64
iy=64
ox=64
oy=64
zoom=1.1
angle=2
convert lena.jpg -distort SRT "$ix,$iy $zoom $angle $ox,$oy" x2.jpg

```

![x2](https://user-images.githubusercontent.com/13458218/140454329-84b1fe3d-cae2-479b-823f-2b26b7f2b85b.jpg)

Or you can rotate and zoom about some point and then move that point to the center of the output. Here the %[fx:w/2],%[fx:h/2] computes the center automatically from the input

```
ix=64
iy=64
zoom=1.1
angle=2
convert lena.jpg -distort SRT "$ix,$iy $zoom $angle %[fx:w/2],%[fx:h/2]" x3.jpg

```

![x3](https://user-images.githubusercontent.com/13458218/140454357-eb33a74e-eec7-485d-9d5e-1f0ce6c9991c.jpg)

For IM 7, change convert to magick.
